File: backend_samples/library/java/AzureAIVisionFaceDeviceAttestation/src/main/java/com/azure/ai/vision/face/deviceattestation/ios/AppAttestParsers.java
package com.azure.ai.vision.face.deviceattestation.ios;

import static com.azure.ai.vision.face.deviceattestation.ios.AppAttestConstants.AUTH_DATA_HEADER_BYTES;
import static com.azure.ai.vision.face.deviceattestation.ios.AppAttestConstants.FLAGS_OFFSET;
import static com.azure.ai.vision.face.deviceattestation.ios.AppAttestConstants.RP_ID_HASH_BYTES;
import static com.azure.ai.vision.face.deviceattestation.ios.AppAttestConstants.SIGN_COUNT_OFFSET;

import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import com.azure.ai.vision.face.deviceattestation.crypto.Base64Utils;
import com.azure.ai.vision.face.deviceattestation.crypto.CertUtils;
import org.bouncycastle.asn1.ASN1OctetString;
import org.bouncycastle.asn1.ASN1Sequence;
import org.bouncycastle.asn1.ASN1TaggedObject;

/** Decoders for the two CBOR payloads Apple's App Attest APIs return. */
public final class AppAttestParsers {

    /** Decoded first 37 bytes of authenticatorData. */
    public record AssertionAuthData(byte[] rpIdHash, int flags, long signCount) {
    }

    /** Decode the base64-CBOR attestation token (DCAppAttestService.attestKey). */
    @SuppressWarnings("unchecked")
    public static AppAttestObject parseAppAttestToken(String tokenB64) {
        byte[] tokenBuf = Base64Utils.decode(tokenB64);
        Object topRaw = Cbor.decode(tokenBuf).value();
        if (!(topRaw instanceof Map)) {
            throw new IllegalArgumentException("top-level CBOR is not a map");
        }
        Map<Object, Object> top = (Map<Object, Object>) topRaw;

        if (!(top.get("fmt") instanceof String fmt)) {
            throw new IllegalArgumentException("missing fmt");
        }
        if (!(top.get("attStmt") instanceof Map)) {
            throw new IllegalArgumentException("missing attStmt");
        }
        Map<Object, Object> attStmt = (Map<Object, Object>) top.get("attStmt");
        if (!(top.get("authData") instanceof byte[] authData)) {
            throw new IllegalArgumentException("missing authData");
        }
        if (!(attStmt.get("x5c") instanceof List<?> x5c) || x5c.size() < 2) {
            throw new IllegalArgumentException("attStmt.x5c must have >= 2 certs");
        }
        if (!(x5c.get(0) instanceof byte[] credCertDer) || !(x5c.get(1) instanceof byte[] intermediateDer)) {
            throw new IllegalArgumentException("attStmt.x5c entries must be byte strings");
        }

        byte[] receipt = attStmt.get("receipt") instanceof byte[] r ? r : null;
        return new AppAttestObject(fmt, authData, credCertDer, intermediateDer,
                receipt != null ? receipt.length : 0, receipt);
    }

    /** Decode the base64-CBOR assertion (DCAppAttestService.generateAssertion). */
    @SuppressWarnings("unchecked")
    public static AppAttestAssertionObject parseAppAttestAssertion(String assertionB64) {
        byte[] buf = Base64Utils.decode(assertionB64);
        Object topRaw = Cbor.decode(buf).value();
        if (!(topRaw instanceof Map)) {
            throw new IllegalArgumentException("assertion top-level CBOR is not a map");
        }
        Map<Object, Object> top = (Map<Object, Object>) topRaw;
        if (!(top.get("signature") instanceof byte[] signature)) {
            throw new IllegalArgumentException("assertion missing signature");
        }
        if (!(top.get("authenticatorData") instanceof byte[] authenticatorData)) {
            throw new IllegalArgumentException("assertion missing authenticatorData");
        }
        if (authenticatorData.length < AUTH_DATA_HEADER_BYTES) {
            throw new IllegalArgumentException("assertion authenticatorData is " + authenticatorData.length + " bytes, < " + AUTH_DATA_HEADER_BYTES);
        }
        return new AppAttestAssertionObject(signature, authenticatorData);
    }

    /** Decode the first 37 bytes: rpIdHash(32) || flags(1) || signCount(4). */
    public static AssertionAuthData parseAssertionAuthData(byte[] authenticatorData) {
        if (authenticatorData.length < AUTH_DATA_HEADER_BYTES) {
            throw new IllegalArgumentException("authenticatorData is " + authenticatorData.length + " bytes, < " + AUTH_DATA_HEADER_BYTES);
        }
        return new AssertionAuthData(
                Arrays.copyOfRange(authenticatorData, 0, RP_ID_HASH_BYTES),
                Byte.toUnsignedInt(authenticatorData[FLAGS_OFFSET]),
                Integer.toUnsignedLong(ByteBuffer.wrap(authenticatorData, SIGN_COUNT_OFFSET, Integer.BYTES).getInt()));
    }

    /**
     * Pull the nonce OCTET STRING out of credCert extension OID
     * 1.2.840.113635.100.8.2.
     */
    public static byte[] extractNonceFromCredCert(byte[] credCertDer) {
        byte[] container = CertUtils.getExtensionValue(credCertDer, AppAttestConstants.NONCE_OID);
        if (container == null) {
            return null;
        }
        try {
            ASN1Sequence sequence = ASN1Sequence.getInstance(container);
            if (sequence.size() != 1) return null;
            ASN1TaggedObject tagged = ASN1TaggedObject.getInstance(sequence.getObjectAt(0));
            if (!tagged.hasContextTag(1) || !tagged.isExplicit()) return null;
            return ASN1OctetString.getInstance(tagged, true).getOctets();
        } catch (RuntimeException exception) {
            return null;
        }
    }

    private AppAttestParsers() {
    }
}

File: backend_samples/library/dotnet/AzureAIVisionFaceDeviceAttestation/src/Ios/AppAttestParsers.cs
using System.Buffers.Binary;
using System.Formats.Asn1;
using Azure.AI.Vision.Face.DeviceAttestation.Crypto;
using static Azure.AI.Vision.Face.DeviceAttestation.Ios.AppAttestConstants;

namespace Azure.AI.Vision.Face.DeviceAttestation.Ios;

/// <summary>Decoded shape of the CBOR attestation object.</summary>
internal sealed class AppAttestObject
{
    public required string Fmt { get; init; }
    public required byte[] AuthData { get; init; }
    public required byte[] CredCertDer { get; init; }
    public required byte[] IntermediateDer { get; init; }
    public int ReceiptLength { get; init; }
    public byte[]? Receipt { get; init; }
}

/// <summary>Decoded shape of the CBOR assertion object.</summary>
internal sealed class AppAttestAssertionObject
{
    public required byte[] Signature { get; init; }
    public required byte[] AuthenticatorData { get; init; }
}

/// <summary>Decoders for the two CBOR payloads Apple's App Attest APIs return.</summary>
internal static class AppAttestParsers
{
    /// <summary>Decode the base64-CBOR attestation token (DCAppAttestService.attestKey).</summary>
    public static AppAttestObject ParseAppAttestToken(string tokenB64)
    {
        var tokenBuf = Base64Utils.Decode(tokenB64);
        var (topRaw, _) = Cbor.Decode(tokenBuf);
        if (topRaw is not Dictionary<object, object?> top)
        {
            throw new FormatException("top-level CBOR is not a map");
        }

        if (MapGet(top, "fmt") is not string fmt)
        {
            throw new FormatException("missing fmt");
        }
        if (MapGet(top, "attStmt") is not Dictionary<object, object?> attStmt)
        {
            throw new FormatException("missing attStmt");
        }
        if (MapGet(top, "authData") is not byte[] authData)
        {
            throw new FormatException("missing authData");
        }
        if (MapGet(attStmt, "x5c") is not List<object?> x5c || x5c.Count < 2)
        {
            throw new FormatException("attStmt.x5c must have >= 2 certs");
        }
        if (x5c[0] is not byte[] credCertDer || x5c[1] is not byte[] intermediateDer)
        {
            throw new FormatException("attStmt.x5c entries must be byte strings");
        }

        var receipt = MapGet(attStmt, "receipt") as byte[];
        return new AppAttestObject
        {
            Fmt = fmt,
            AuthData = authData,
            CredCertDer = credCertDer,
            IntermediateDer = intermediateDer,
            ReceiptLength = receipt?.Length ?? 0,
            Receipt = receipt,
        };
    }

    /// <summary>Decode the base64-CBOR assertion (DCAppAttestService.generateAssertion).</summary>
    public static AppAttestAssertionObject ParseAppAttestAssertion(string assertionB64)
    {
        var buf = Base64Utils.Decode(assertionB64);
        var (topRaw, _) = Cbor.Decode(buf);
        if (topRaw is not Dictionary<object, object?> top)
        {
            throw new FormatException("assertion top-level CBOR is not a map");
        }
        if (MapGet(top, "signature") is not byte[] signature)
        {
            throw new FormatException("assertion missing signature");
        }
        if (MapGet(top, "authenticatorData") is not byte[] authenticatorData)
        {
            throw new FormatException("assertion missing authenticatorData");
        }
        if (authenticatorData.Length < AuthDataHeaderBytes)
        {
            throw new FormatException($"assertion authenticatorData is {authenticatorData.Length} bytes, < {AuthDataHeaderBytes}");
        }
        return new AppAttestAssertionObject { Signature = signature, AuthenticatorData = authenticatorData };
    }

    /// <summary>Decode the first 37 bytes: rpIdHash(32) || flags(1) || signCount(4).</summary>
    public static (byte[] RpIdHash, int Flags, long SignCount) ParseAssertionAuthData(byte[] authenticatorData)
    {
        if (authenticatorData.Length < AuthDataHeaderBytes)
        {
            throw new FormatException($"authenticatorData is {authenticatorData.Length} bytes, < {AuthDataHeaderBytes}");
        }
        return (authenticatorData[..RpIdHashBytes], authenticatorData[FlagsOffset],
            BinaryPrimitives.ReadUInt32BigEndian(authenticatorData.AsSpan(SignCountOffset)));
    }

    /// <summary>
    /// Pull the nonce OCTET STRING out of credCert extension OID
    /// 1.2.840.113635.100.8.2. Structure:
    ///   Extension ::= SEQUENCE { extnID OID, extnValue OCTET STRING }
    ///   extnValue (DER) = SEQUENCE { [1] EXPLICIT OCTET STRING nonce }
    /// </summary>
    public static byte[]? ExtractNonceFromCredCert(byte[] credCertDer)
    {
        var container = CertUtils.GetExtensionValue(credCertDer, AppAttestConstants.NonceOid);
        if (container is null)
        {
            return null;
        }
        try
        {
            var reader = new AsnReader(container, AsnEncodingRules.DER);
            var sequence = reader.ReadSequence();
            var tagged = sequence.ReadSequence(new Asn1Tag(TagClass.ContextSpecific, 1, isConstructed: true));
            var nonce = tagged.ReadOctetString();
            tagged.ThrowIfNotEmpty();
            sequence.ThrowIfNotEmpty();
            reader.ThrowIfNotEmpty();
            return nonce;
        }
        catch (AsnContentException)
        {
            return null;
        }
    }

    private static object? MapGet(Dictionary<object, object?> map, string key)
        => map.TryGetValue(key, out var value) ? value : null;

}
